src/select/Select.ts

```diff
@@ -0,0 +1,411 @@
+const keys = {
+	escape: 27,
+	enter: 13,
+	space: 32,
+	up: 38,
+	down: 40,
+	home: 36,
+	end: 35
+};
+
+export const SelectBase = ThemeableMixin(WidgetBase);
+
+@theme(css)
+export default class Select extends SelectBase<SelectProperties> {
+	private _focusedIndex = 0;
+	private _ignoreBlur = false;
+	private _open = false;
+	private _selectId: string = uuid();
+
+	private _onBlur (event: FocusEvent) { this.properties.onBlur && this.properties.onBlur(event); }
+	private _onClick (event: MouseEvent) { this.properties.onClick && this.properties.onClick(event); }
+	private _onFocus (event: FocusEvent) { this.properties.onFocus && this.properties.onFocus(event); }
+	private _onKeyDown (event: KeyboardEvent) { this.properties.onKeyDown && this.properties.onKeyDown(event); }
+
+	// native select events
+	private _onNativeChange (event: Event) {
+		const { options = [] } = this.properties;
+		const option = options.filter((option: SelectOption) => option.value === (<HTMLInputElement> event.target).value)[0];
+		this.properties.onChange && this.properties.onChange(option);
+	}
+
+	// custom select events
+	private _openSelect() {
+		this._open = true;
+		this._ignoreBlur = false;
+		this._focusedIndex = this._focusedIndex || 0;
+		this.invalidate();
+	}
+	private _closeSelect() {
+		this._open = false;
+		this.invalidate();
+	}
+
+	private _onTriggerClick(event: MouseEvent) {
+		this.properties.onClick && this.properties.onClick(event);
+
+		this._open ? this._closeSelect() : this._openSelect();
+	}
+
+	private _onTriggerBlur(event: FocusEvent) {
+		if (!this._ignoreBlur) {
+			this.properties.onBlur && this.properties.onBlur(event);
+			this._closeSelect();
+		}
+	}
+
+	private _onOptionMouseDown() {
+		this._ignoreBlur = true;
+	}
+
+	private _onOptionClick(event: MouseEvent) {
+		const {
+			options = [],
+			onChange,
+			onClick
+		} = this.properties;
+
+		onClick && onClick(event);
+
+		// find parent option node, and get index
+		let optionNode = <HTMLElement> event.target;
+		while (!optionNode.hasAttribute('data-dojo-index') && optionNode.parentElement) {
+			optionNode = optionNode.parentElement;
+		}
+
+		const index = optionNode.getAttribute('data-dojo-index');
+		if (index) {
+			const option = options[parseInt(index, 10)];
+
+			if (option && !option.disabled) {
+				this._focusedIndex = parseInt(index, 10);
+				onChange && onChange(option);
+			}
+		}
+	}
+
+	private _onListboxKeyDown(event: KeyboardEvent) {
+		this.properties.onKeyDown && this.properties.onKeyDown(event);
+
+		const {
+			options = [],
+			multiple,
+			onChange
+		} = this.properties;
+		const { _focusedIndex } = this;
+
+		switch (event.which) {
+			case keys.enter:
+				!options[_focusedIndex].disabled && onChange && onChange(options[_focusedIndex]);
+				break;
+			case keys.space:
+				!options[_focusedIndex].disabled && onChange && onChange(options[_focusedIndex]);
+				break;
+			case keys.escape:
+				this._closeSelect();
+				break;
+			case keys.down:
```

Comment:
Can a `disabled` option be focused using the keyboard? I didn't see any logic to find the next enabled option if the currently-focused option is disabled. Curious because I had to do something like that for TabPane.